GraalVM issue with ArrayList serialization

### Describe the bug

When trying to serialize an `ArrayList` in native, we get the following error:

```
2021-08-27 07:00:01,618 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-0) HTTP Request to /hello/serlist failed, error id: 4735770b-c272-499f-ad87-5b70a6a2ca6c-1: org.jboss.resteasy.spi.UnhandledException: java.io.InvalidClassException: java.util.ArrayList; no valid constructor
        at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
        at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
        at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
        at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:138)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:93)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:536)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:829)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:553)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
Caused by: java.io.InvalidClassException: java.util.ArrayList; no valid constructor
        at java.io.ObjectStreamClass$ExceptionInfo.newInvalidClassException(ObjectStreamClass.java:159)
        at java.io.ObjectStreamClass.checkDeserialize(ObjectStreamClass.java:875)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2170)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1679)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:493)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:451)
        at org.acme.GreetingResource.serlist(GreetingResource.java:69)
        at java.lang.reflect.Method.invoke(Method.java:566)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
        at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
        at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:408)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:69)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
        ... 17 more

```

Looks similar to https://github.com/micronaut-projects/micronaut-redis/issues/194

### Expected behavior

The list should be able to be serialized and un-serialized.

### Actual behavior

It fails with the above exception.

### How to Reproduce?

Reproducer:

- Create a sample JAXRS application
- Add the following code:

```
    @GET
    @Produces(MediaType.TEXT_PLAIN)
    @Path("/serstring")
    public String serstring() throws IOException, ClassNotFoundException {

        byte[] bytes = null;

        try (
                ByteArrayOutputStream baos = new ByteArrayOutputStream();
                ObjectOutputStream oos = new ObjectOutputStream(baos)) {

            oos.writeObject("Hello RESTEasy");
            oos.flush();
            bytes = baos.toByteArray();
        }

        try (ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
                ObjectInputStream ois = new ObjectInputStream(bais)) {
            return (String) ois.readObject();
        }
    }

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    @Path("/serlist")
    public String serlist() throws IOException, ClassNotFoundException {

        byte[] bytes = null;

        try (
                ByteArrayOutputStream baos = new ByteArrayOutputStream();
                ObjectOutputStream oos = new ObjectOutputStream(baos)) {

            List<String> list = new ArrayList<>();
            list.add("Hello RESTEasy");

            oos.writeObject(list);
            oos.flush();
            bytes = baos.toByteArray();
        }

        try (ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
                ObjectInputStream ois = new ObjectInputStream(bais)) {
            
            return ois.readObject().toString();
        }
    }
```

- do a native build
- start the server
- execute `curl http://localhost:8080/hello/serstring` => `Hello RESTEasy`
- execute `curl http://localhost:8080/hello/serlist` => exception

### Output of `uname -a` or `ver`

windows

### Output of `java -version`

openjdk 11.0.1-redhat 2018-10-16 LTS

### GraalVM version (if different from Java)

Running Quarkus native-image plugin on GraalVM 21.1.0 Java 11 CE (Java Version 11.0.11+8-jvmci-21.1-b05)

### Quarkus version or git rev

2.1.4.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)

### Additional information

`<quarkus.native.container-build>true</quarkus.native.container-build>`

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java
@@ -567,8 +567,6 @@ public class NativeImageAutoFeatureStep {
                 greaterThan21_3.loadNull(), registerArgs);
         greaterThan21_3.returnValue(null);
 
-        ResultHandle objectClass = tc.invokeStaticMethod(forNameMethodDescriptor, tc.load("java.lang.Object"),
-                tc.load(false), tccl);
         ResultHandle serializationRegistryClass = tc.invokeStaticMethod(forNameMethodDescriptor,
                 tc.load("com.oracle.svm.core.jdk.serialize.SerializationRegistry"),
                 tc.load(false), tccl);
@@ -670,7 +668,7 @@ public class NativeImageAutoFeatureStep {
                 serializationSupport, clazz, newSerializationConstructorClass, accessor);
         ResultHandle addReflectionsArgs2 = tc.newArray(Class.class, tc.load(2));
         tc.writeArrayValue(addReflectionsArgs2, 0, clazz);
-        tc.writeArrayValue(addReflectionsArgs2, 1, objectClass);
+        tc.writeArrayValue(addReflectionsArgs2, 1, newSerializationConstructorClass);
         tc.invokeVirtualMethod(invokeMethodDescriptor, addReflectionsLookupMethod, tc.loadNull(), addReflectionsArgs2);
 
         addSerializationForClass.returnValue(null);

--- a/integration-tests/main/src/main/java/io/quarkus/it/corestuff/SerializationTestEndpoint.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/corestuff/SerializationTestEndpoint.java
@@ -6,6 +6,8 @@ import java.io.IOException;
 import java.io.ObjectInputStream;
 import java.io.ObjectOutputStream;
 import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.List;
 
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
@@ -42,6 +44,7 @@ public class SerializationTestEndpoint extends HttpServlet {
             ExternalizablePerson ep = new ExternalizablePerson();
             ep.setName("Sheldon 2.0");
             instance.setExternalizablePerson(ep);
+            instance.setList(new ArrayList<>(List.of("Hello World!")));
             ByteArrayOutputStream out = new ByteArrayOutputStream();
             ObjectOutputStream os = new ObjectOutputStream(out);
             os.writeObject(instance);
@@ -49,7 +52,8 @@ public class SerializationTestEndpoint extends HttpServlet {
             ObjectInputStream is = new ObjectInputStream(bais);
             SomeSerializationObject result = (SomeSerializationObject) is.readObject();
             if (result.getPerson().getName().equals("Sheldon")
-                    && result.getExternalizablePerson().getName().equals("Sheldon 2.0")) {
+                    && result.getExternalizablePerson().getName().equals("Sheldon 2.0")
+                    && result.getList().toString().equals("[Hello World!]")) {
                 resp.getWriter().write("OK");
             } else {
                 reportException("Serialized output differs from input", null, resp);

--- a/integration-tests/main/src/main/java/io/quarkus/it/corestuff/serialization/SerializationConfig.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/corestuff/serialization/SerializationConfig.java
@@ -1,8 +1,11 @@
 package io.quarkus.it.corestuff.serialization;
 
+import java.util.ArrayList;
+import java.util.List;
+
 import io.quarkus.runtime.annotations.RegisterForReflection;
 
-@RegisterForReflection(targets = String.class, serialization = true)
+@RegisterForReflection(targets = { List.class, ArrayList.class, String.class }, serialization = true)
 public class SerializationConfig {
 
 }

--- a/integration-tests/main/src/main/java/io/quarkus/it/corestuff/serialization/SomeSerializationObject.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/corestuff/serialization/SomeSerializationObject.java
@@ -1,6 +1,7 @@
 package io.quarkus.it.corestuff.serialization;
 
 import java.io.Serializable;
+import java.util.List;
 
 import io.quarkus.runtime.annotations.RegisterForReflection;
 
@@ -9,6 +10,7 @@ public class SomeSerializationObject implements Serializable {
 
     private Person person;
     private ExternalizablePerson externalizablePerson;
+    private List<String> list;
 
     public Person getPerson() {
         return person;
@@ -24,5 +26,13 @@ public class SomeSerializationObject implements Serializable {
 
     public void setExternalizablePerson(ExternalizablePerson externalizablePerson) {
         this.externalizablePerson = externalizablePerson;
+    }
+
+    public List<String> getList() {
+        return list;
+    }
+
+    public void setList(List<String> list) {
+        this.list = list;
     }
 }

